text/0000-slatepack.md

```diff
@@ -0,0 +1,391 @@
+      - `message key` is 128-bit master message key generated by each recipient
+      - `encrypt[key](plaintext)` is ChaCha20-Poly1305 from RFC 7539 with a zero nonce.
+      - `HKDF[salt, label](key)` is 32 bytes of HKDF from RFC 5869 with SHA-256.
+      - `salt` is `X25519(ephemeral secret, basepoint) || public key`
+      - `label` is "slatepack_v0.1"
+
+- `mac`
+  - A Message Authentication Code is generated on the `header` data which is all of the header values concatenated
+    - `encode(HMAC[HKDF["", "header"](message key)](header))`
+      - `encode(data)` is canonical base64 from RFC 4648 without padding.
+      - `HMAC[key](message)` is HMAC from RFC 2104 with SHA-256.
+      - `HKDF[salt, label](key)` is 32 bytes of HKDF from RFC 5869 with SHA-256.
+
+#### Payload (age Enryption)
+
+A binary serialized slate is encrypted according to the age encryption specification. The steps taken here follow age as closely as possible to avoid losing any security properties. An fairly well-review [age library in rust](https://github.com/str4d/rage/tree/master/age) is available to use for implementation.
+
+Any deviations in SlatePack encryption from the exact cryptography steps and decisions made in age are unintentional and should be corrected unless they are explicitly stated as a deviation from the cryptography decisions made by age.
+
+The expected deviations from the age specification are in the formatting and structure of the contents of the file output data (as opposed to the encryption technique itself or the actual binary encrypted payload).
+
+- `payload`
+  - `nonce || STREAM[HKDF[nonce, "payload"](message key)](plaintext)`
+    - `nonce` is `random(16)`
+    - `message key` is 128-bit master message key for each recipient
+    - `STREAM` is `ChaCha20-Poly1305` in 64KiB chunks and a nonce structure of 11 bytes of big endian counter, and 1 byte of last block flag (0x00 / 0x01)
+    - `HKDF[salt, label](key)` is 32 bytes of HKDF from RFC 5869 with SHA-256
+
+### Armor
+
+The payload that will be armored is a binary serialized `SlatePack` JSON object. Armor is `Framing` wrapped around a `SimpleBase58Check` encoded `Payload`.
+
+#### Framing
+
+Armor uses specific `Headers`, `Footers` and `Periods` as `Framing` to contain its `Payload`.
+
+- `Header`
+  - Supported Headers:
+    - `BEGINSLATEPACK`
+    - `ENDSLATEPACK X/Y`
+      - `Y` is the total number of message parts for a given multipart message and `X` is the number of the current partial message
+      - Only used if there is more than one message part
+  - Regex: `^[>\n\r\t ]*BEGINSLATEPACK[>\n\r\t ]*([1-9]\/[1-9])*$`
+
+
+- `Footer`
+  - Supported Footers
+    - `ENDSLATEPACK`
+    - `ENDSLATEPACK X/Y`
+      - `Y` is the total number of message parts for a given multipart message and `X` is the number of the current partial message
+      - Only used if there is more than one message part
+  - Regex: `^[>\n\r\t ]*ENDSLATEPACK[>\n\r\t ]*([1-9]\/[1-9])*$`
```

Comment:
Since this is to be copy-pasted, I'm not sure it makes sense to support any \n\r\t before or after ENDSLATEPACK. It's better to be strict here, at least when generating. Implementations can be loose with their validation if they so prefer, but should never generate superfluous spaces, tabs, etc.

For this, I would recommend just `^ENDSLATEPACK[ ]?([1-9]\/[1-9])?$`

Reply:
On second thought, neither your begin nor end regex validates for the example:
```
BEGINSLATEPACK. 4H1qx1wHe668tFW yC2gfL8PPd8kSgv
pcXQhyRkHbyKHZg GN75o7uWoT3dkib R2tj1fFGN2FoRLY
GWmtgsneoXf7N4D uVWuyZSamPhfF1u AHRaYWvhF7jQvKx
wNJAc7qmVm9JVcm NJLEw4k5BU7jY6S eb. ENDSLATEPACK.
```

Reply:
The regex work is mostly borrowed from https://saltpack.org/armoring as well. They were briefly tested with `regex` in rust but should certainly be reviewed to make sure they are exactly we want for an implementation. I expect it will need to be further refined as you have done :+1:

Reply:
> Since this is to be copy-pasted, I'm not sure it makes sense to support any \n\r\t before or after ENDSLATEPACK. It's better to be strict here, at least when generating. Implementations can be loose with their validation if they so prefer, but should never generate superfluous spaces, tabs, etc.

Same applies as my comment here: https://github.com/mimblewimble/grin-rfcs/pull/55#discussion_r424946340

We should not make any assumptions about _how_ slatepacks will be shared, they could be forwarded via emailed, IM, and a myriad of different ways. So it's less about implementations generating superfluous characters, and more about extra tab, space, or line break characters that may be oblivious to the end user, that get sneakily inserted as part of the _communication channels_ they use that can lead to unexpected failure.

Reply:
> On second thought, neither your begin nor end regex validates for the example:

Yes, this is due to the definitions of both the header and footer regexes:
``` regex
^[>\n\r\t ]*BEGINSLATEPACK[>\n\r\t ]*([1-9]\/[1-9])*$
^[>\n\r\t ]*ENDSLATEPACK[>\n\r\t ]*([1-9]\/[1-9])*$
```

`$` here asserts position at the end of the line. Removing `$` makes the regex validate. This appears in the saltpack definitions as well, and looks like a bug to me.

Reply:
Right, I was referring to the `$` and also the `^` which refers to the beginning of a line, which isn't correct for ENDSLATEPACK.

Reply:
Good catch, `^` doesn't belong there either, not for ENDSLATEPACK, and possibly not for  BEGINSLATEPACK either? Should we really be making assumptions about whether the message always begins on a new line?